File: api/src/main/java/com/wbrawner/twigs/category/Category.java
package com.wbrawner.twigs.category;

import com.wbrawner.twigs.budget.Budget;
import jakarta.persistence.*;

import static com.wbrawner.twigs.Utils.randomId;

@Entity
public class Category implements Comparable<Category> {
    @Id
    private final String id = randomId();
    private String title;
    private String description;
    private long amount;
    @JoinColumn(nullable = false)
    @ManyToOne
    private Budget budget;
    private boolean expense;
    @Column(nullable = false, columnDefinition = "boolean default false")
    private boolean archived;

    public Category() {
        this(null, null, 0L, null, true);
    }

    public Category(
            String title,
            String description,
            Long amount,
            Budget budget,
            boolean expense
    ) {
        this.title = title;
        this.description = description;
        this.amount = amount;
        this.budget = budget;
        this.expense = expense;
    }

    @Override
    public int compareTo(Category other) {
        return title.compareTo(other.title);
    }

    public String getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public long getAmount() {
        return amount;
    }

    public void setAmount(long amount) {
        this.amount = amount;
    }

    public Budget getBudget() {
        return budget;
    }

    public void setBudget(Budget budget) {
        this.budget = budget;
    }

    public boolean isExpense() {
        return expense;
    }

    public void setExpense(boolean expense) {
        this.expense = expense;
    }

    public boolean isArchived() {
        return archived;
    }

    public void setArchived(boolean archived) {
        this.archived = archived;
    }
}

File: api/src/main/kotlin/com/wbrawner/twigs/server/category/Category.kt
package com.wbrawner.twigs.server.category

import com.wbrawner.twigs.server.budget.Budget
import com.wbrawner.twigs.server.randomString
import javax.persistence.*

@Entity
data class Category(
    @Id
    val id: String = randomString(),
    var title: String= "",
    var description: String? = null,
    var amount: Long = 0L,
    @field:ManyToOne
    @field:JoinColumn(nullable = false)
    var budget: Budget? = null,
    var expense: Boolean = true,
    @field:Column(nullable = false, columnDefinition = "boolean default false")
    var archived: Boolean = false
)

data class NewCategoryRequest(
    val title: String,
    val description: String? = null,
    val amount: Long,
    val budgetId: String,
    val expense: Boolean
)

data class UpdateCategoryRequest(
    val title: String? = null,
    val description: String? = null,
    val amount: Long? = null,
    val expense: Boolean? = null,
    val archived: Boolean? = null
)

data class CategoryResponse(
    val id: String,
    val title: String,
    val description: String?,
    val amount: Long,
    val budgetId: String,
    val isExpense: Boolean,
    val isArchived: Boolean
) {
    constructor(category: Category) : this(
        category.id,
        category.title,
        category.description,
        category.amount,
        category.budget!!.id,
        category.expense,
        category.archived
    )
}

data class CategoryBalanceResponse(val id: String, val balance: Long)